### PR TITLE
Upgrade Wildfly to 18.0.1.Final

### PR DESCRIPTION
**Pull request by mareknovotny**
Change-Id: Ibfc9ba78c40b5d6b72eda7d7ff437da4be691de8

https://issues.redhat.com/browse/JBPM-9044

List of PRs:
* this PR #1203 
* kiegroup/appformer/pull/910
* kiegroup/jbpm/pull/1601
* kiegroup/droolsjbpm-integration/pull/2033
* kiegroup/jbpm-work-items/pull/125
* kiegroup/kie-wb-common/pull/3273
* kiegroup/drools-wb/pull/1349
* kiegroup/jbpm-wb/pull/1429
* kiegroup/optaplanner-wb/pull/368
* kiegroup/kie-wb-distributions/pull/1022

### CODE REVIEW

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by sutaakar** [comment]
@mareknovotny The Kie server tests need these changes (in order for JMS Initial context to be initialized properly in tests) - https://github.com/sutaakar/droolsjbpm-integration/commit/27aef967c85b136c5dae437cdeee5f6b9599d03a

With this change the smoke tests seem to pass.

**Comment by mareknovotny** [comment]
@sutaakar  do you know, if Is there a diff between https://codehaus-cargo.github.io/cargo/WildFly+18.x.html and our current 14x cargo container setup?

**Comment by sutaakar** [comment]
There should be no difference. We should update the container id for Cargo to keep consistency, but functionality wise it will be still the same.

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
@sutaakar @MarianMacik any help with failed kie-camel integration tests ? All failed with HTTP 404

```
Error Message

HTTP 404 Not Found

Stacktrace

javax.ws.rs.NotFoundException: HTTP 404 Not Found
	at org.kie.camel.container.integration.tests.ProcessEngineIntegrationTest.testStartProcess(ProcessEngineIntegrationTest.java:53)
```

**Comment by jiripetrlik** [comment]
@mareknovotny 
According to wildfly-camel page [1] Wildfly 18.01 is aligned with camel patch 11.0.0. I would suggest to change property "version.org.wildfly.camel" to "11.0.0". However, I did not test this solution. Give me some time to test whether it is ok.
[1] https://github.com/wildfly-extras/wildfly-camel/releases

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
so we are blocked now on camel and its unsupported combination version 2.x and Wildfly 18+

**Comment by mareknovotny** [comment]
blocked now by wildfly-extras/wildfly-camel/issues/2962

**Comment by mbiarnes** [comment]
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1203/files#diff-600376dffeb79835ede4a0b285078036R261 should be upgraded to version 10.1.2.Final to align with EAP 7.3.0

**Comment by mareknovotny** [comment]
@mbiarnes ok, I took here WF deps versions, so it could be that EAP is upgrading some of them as follow up changes too, no big deal here, but I can upgrade that

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
Hey @mareknovotny 

Thanks for the updates. Just a couple of comments:

* I understand you already did the job of reviewing the server side dependency updates necessary for wildfly 18.x, I mean, I'm not sure about each concrete version update but it looks coherent for me, it looks good to me. Although that said, we should test some our showcases (being used during development) before finally merging this, from both command line and also from our IDE, to ensure no issue at runtime will prevent us for moving forward. Can we assign somebody from QE to do some smoke tests on some showcase for this? WDYT?

* I think it's maybe missing to also update the Wifldly version being used by our Errai/GWT server during development mode. For example, look at the `gwt-maven-plugin` configuration for this showcase in Stunner ([see here](https://github.com/kiegroup/kie-wb-common/blob/7bc7b538e138162809db65ce0cd441a76079f203/kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/pom.xml#L910)), it relies on some version.org.jboss.errai.wildfly but it looks not being updated in the parent or any other pom. so maybe that showcases will be still running with the old Wildfly versions, instead of the ones assembled commandly line wich are being adapted to latest version. Can you also please if that Errai wildfly version update is also necessary, and everything works after this change?

CC @porcelli @ederign @manstis @cristianonicolai Are we missing something else for doing the upgrade? Please some quick check from you would also help a lot.

Thank u all guys, in advance!

**Comment by mareknovotny** [comment]
+1 to test showcases and IDE @romartin 
The Errai wildfly container was not upgraded as it seems was not upgraded for long time. I recall when Max created the Errai WF container it was for solving the issue with GWT SDM which needed a workaround to load some modules. Maybe we could avoid to use that special container at all now. Or if that is just for GWT SDM it is not so important, but I am here not expert on GWT ;) so i would need a help from somebody to try that and possibly do new updated version in Errai. https://github.com/errai/errai/tree/master/errai-cdi/errai-wildfly-dist

**Comment by porcelli** [comment]
do you mind put this on hold until we release Kogito Dev Preview?

**Comment by romartin** [comment]
@mareknovotny yeah TBH I also don't know "how" necessary is the Errai Wildfly launcher at this point, although we still depen on GWT and most of the showcases also rely on this. So we should give a try to this, but as @porcelli suggests, maybe it can be after next kogito release? I can help too.

**Comment by mareknovotny** [comment]
we don't need to wait for Errai Wildfly container check and upgrade if needed imho, we can do it as second step as GWT SDM is just developer environment, isn't it? @porcelli @romartin

**Comment by porcelli** [comment]
@mareknovotny GWT setup sometimes is very trick, and I'm not comfortable to merge this now and follow on dev setup later. Can't we put this on hold, until at least have Kogito DevPreview released?

**Comment by jstastny-cz** [comment]
Make sure we don't merge this unless Full downstream build succeeds. Once it does, still before merging, QE will run test suites with the artifacts. Don't merge this before that.

**Comment by mareknovotny** [comment]
I am not merging anything until it is green @jstastny-cz @porcelli 
On the other side what I don't want is waiting on something which is not directly blocked by this changeset, it waited here for more than month due camel wildfly integration so if all would be green and the SDM mode won't be ready it is the independent thing to this. We must upgrade to WF 18 or higher

**Comment by porcelli** [comment]
@mareknovotny if the dev mode doesn't work, it's a blocker to this change. the problem is that today we are all focused on Kogito DevPreview.

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by adrielparedes** [comment]
@mareknovotny did you have to upgrade any infinispan library? Can we possibly have a problem with duplicated libraries between ispn version we use and the version wildfly18 includes? @ederign just in case this could affect HA. Since we are still using eap7 for prod shouldn’t be a problem but we need to double check this for community release

**Comment by mareknovotny** [comment]
@adrielparedes i don't think it should be a problem, ISPN in WF is supported only for internal deps like  second level cache in Hibernate and clustering itself, but surely we need to check that too

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Adding also upgrade of extra-inforcer-rules to be more valid https://github.com/mojohaus/extra-enforcer-rules/issues/46

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Failures in kie-wb-common aka stunner?
```    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.testIntOverflowExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.testNegativeDurationReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.testIntMaxExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.testIsTooBigDurationReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.testEmptyReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test10DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test1DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test2DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test3DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test4DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.ReassignmentValueValidatorTest.test5DigExpiresAtReassignmentValue
    org.kie.workbench.common.stunner.bpmn.forms.validation.timerEditor.TimerSettingsValueValidatorTest.testValidateTimeDuration
    org.kie.workbench.common.stunner.bpmn.forms.validation.timerEditor.TimerSettingsValueValidatorTest.testValidateCronTimeCycle
    org.kie.workbench.common.stunner.bpmn.forms.validation.timerEditor.TimerSettingsValueValidatorTest.testValidateTimeDate
    org.kie.workbench.common.stunner.bpmn.forms.validation.timerEditor.TimerSettingsValueValidatorTest.testValidateISOTimeCycle
```

and @jomarko one additional weird error to connect chinese npm registry ? from https://github.com/kiegroup/kie-wb-common/blame/master/kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/src/test/java/org/kie/workbench/common/dmn/showcase/client/selenium/DMNDesignerKogitoSeleniumIT.java#L131 
`org.kie.workbench.common.dmn.showcase.client.selenium.DMNDesignerKogitoSeleniumIT.org.kie.workbench.common.dmn.showcase.client.selenium.DMNDesignerKogitoSeleniumIT`

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
random flaky test failure. reported as https://issues.redhat.com/browse/AF-2469

**Comment by tomasdavidorg** [comment]
Hi @mareknovotny ,
.I tested the built wars with EAP 7.3.0 and I got this error:
```
Caused by: java.lang.NoSuchMethodError: org.kie.server.api.marshalling.json.JSONMarshaller$2.idResolver(Lcom/fasterxml/jackson/databind/cfg/MapperConfig;Lcom/fasterxml/jackson/databind/JavaType;Ljava/util/Collection;ZZ)Lcom/fasterxml/jackson/databind/jsontype/TypeIdResolver;
	at org.kie.server.api.marshalling.json.JSONMarshaller$2.buildTypeDeserializer(JSONMarshaller.java:245)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.findTypeDeserializer(BasicDeserializerFactory.java:1599)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.findPropertyTypeDeserializer(BasicDeserializerFactory.java:1739)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.resolveMemberAndTypeAnnotations(BasicDeserializerFactory.java:2107)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.constructSettableProperty(BeanDeserializerFactory.java:799)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.addBeanProps(BeanDeserializerFactory.java:522)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:234)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:143)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:411)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:443)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:183)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:27)
	at com.fasterxml.jackson.databind.DeserializationContext.handlePrimaryContextualization(DeserializationContext.java:650)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:484)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:293)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:476)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4389)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4198)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3205)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3173)
	at org.kie.server.api.marshalling.json.JSONMarshaller.unmarshall(JSONMarshaller.java:330)
	at org.kie.server.controller.rest.ControllerUtils.unmarshal(ControllerUtils.java:183)
	at org.kie.server.controller.rest.RestKieServerControllerImpl.connectKieServer(RestKieServerControllerImpl.java:63)
	at org.kie.server.controller.rest.RestKieServerControllerImpl$Proxy$_$$_WeldClientProxy.connectKieServer(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:138)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:517)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:406)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$0(ResourceMethodInvoker.java:370)
	at org.jboss.resteasy.core.interception.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:356)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:372)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:344)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:317)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:440)
	... 74 more
```
It looks like the problem is the jackson-databind library. I checked the library versions and:

- EAP 7.3.0 has jackson-databind-2.10.0
- WildFly 18.0.1 jackson-databind-2.9.10.1

The `idResolver` method is different in the jars.

**Comment by mareknovotny** [comment]
@tomasdavidorg yes, that is correct, i would not try to use EAP 7.3 for this PR, we have phased upgrades 1. Wildfly and 2. EAP for not matching deps

**Comment by tomasdavidorg** [comment]
Ok thanks. I ran some tests (mostly business central tests, not kie-server) and it looks good to me.

**Comment by mareknovotny** [comment]
@ederign @porcelli, @romartin, @manstis, @yurloc, @bxf12315, @Rikkola, @caponetto, @jesuino  we have approval here from QE, now we should test briefly GWT SDM if it is not broken before merging this set of changes. Could you try it briefly? I don't expect breakage as the Errai Wildfly dist which is used there wasn't upgraded, but just for sure.

**Comment by bxf12315** [comment]
> @ederign @porcelli, @romartin, @manstis, @yurloc, @bxf12315, @Rikkola, @caponetto, @jesuino we have approval here from QE, now we should test briefly GWT SDM if it is not broken before merging this set of changes. Could you try it briefly? I don't expect breakage as the Errai Wildfly dist which is used there wasn't upgraded, but just for sure.
@mareknovotny Base on the PR, I test GWT SDM in jbpm-wb, it's working.

**Comment by ederign** [comment]
Hi @bxf12315 I hope so, but could you please check which version you are running dev mode? Could you please check the target dir on your showcase? Probably you will see: wildfly-14.0.1.Final

For historical reasons, for SDM we pack a wf "special" version created on errai. It's really important to keep both runtime (updated on this helpful PR from Marek) and also SDM (@adrielparedes will work on this soonish) aligned in order to prevent surprises in production due to the libraries version mismatch.

I gently ask to hold this PR until @adrielparedes gives his review. 

(/cc @mareknovotny @dgutierr @porcelli )

**Comment by yurloc** [comment]
@mareknovotny No https://github.com/kiegroup/optaplanner-wb PR?

**Comment by mareknovotny** [comment]
@yurloc no, occurence of wildlfy version property or similar need to change there anything from my findings and testing

File: pom.xml [hunk 1]
```diff
@@ -258,9 +258,9 @@
     <version.org.subethamail>1.2</version.org.subethamail>
     <version.org.subethamail.subethasmtp>3.1.6</version.org.subethamail.subethasmtp>
     <version.org.webjars.swagger-ui>2.2.10</version.org.webjars.swagger-ui>
-    <version.org.wildfly>14.0.1.Final</version.org.wildfly>
-    <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
-    <version.org.wildfly.core>6.0.2.Final</version.org.wildfly.core>
+    <version.org.wildfly>18.0.1.Final</version.org.wildfly>
```
Comment (jomarko) [inline pom.xml:261]: @mareknovotny sorry, can not find better place to comment.

Do we still use `version.org.wildfly.gwt.sdm` ? I found used just in `uberfire-archetype/uberfire-project-archetype` where it is even downgraded to 11.0.

To me seems worth to remove it with this PR to facilitate future upgrades of wildfly. What is your opinion? Should be investigated/done separately?

Comment (mareknovotny) [reply pom.xml:261]: we unfortunately use it for showcases and GWT SDM developer environment, I was asking @adrielparedes if we can drop this special Errai WF distro now. That is work related to https://issues.redhat.com/browse/AF-2483

See original creation at https://github.com/errai/errai/commit/890b0b810f7fd269f92e2d5782a7a57293688cd1

Comment (jomarko) [reply pom.xml:261]: ok my concern was we have now `version.org.wildfly.gwt.sdm` and `version.org.jboss.errai.wildfly` for the same use case, but from your comment it seems they have different purpose

Comment (mareknovotny) [reply pom.xml:261]: @jomarko no, you are right I think they are used for the same purpose. My comment was more about if we don't need to use that special distro dependency, we can eliminate it completely ;)

**Comment by mareknovotny** [comment]
rebased due git merge conflict in kie-parent

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by adrielparedes** [comment]
Sorry guys I take some time to answer, I was doing some research. If you need I can do a special version of errai with wildfly18. I'm working on those errai modification to have SDM working with that version.

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
I rebased and added all changes for Errai WF distro. Right now all PRs are green and here are retriggered FDB as previous was failing in 1 test  **org.kie.hacep.LocalStorageKieSessionTest.fireUntilHaltTest** which timed out.

**Comment by mbiarnes** [comment]
Jenkins execute full downstream build

**Pull request merged**